D2GC: Add D2 graph coloring example

william76 opened the pull request:
Relates to #413 

Adds a distance-2 graph coloring example into `examples/graph`

This is a stripped-down version of the D2GC code from the `perf_test` directory and should be mostly self contained.

srajama1 commented on example/graph/KokkosKernels_CrsMatrix.hpp:
```diff
@@ -0,0 +1,150 @@
+/*
```
Strange, why don't we use our CRS Matrix in the example and have another one ?

srajama1 commented on example/graph/KokkosKernels_Example_Distance2GraphColor.cpp:
```diff
@@ -0,0 +1,577 @@
+    switch(algorithm)
+    {
+        case 1:
+            kh.create_distance2_graph_coloring_handle(COLORING_D2_MATRIX_SQUARED);
```
I think some odd tabs are there in this file.

srajama1 commented on example/graph/KokkosKernels_Example_Distance2GraphColor.cpp:
```diff
@@ -0,0 +1,577 @@
+    // ------------------------------------------
+    if(0 != params.output_histogram)
+    {
+        KokkosGraph::Impl::graph_print_distance2_color_histogram(&kh, 
```
This may not be needed in the example.

srajama1 commented on example/graph/KokkosKernels_StaticCrsGraph.hpp:
```diff
@@ -0,0 +1,130 @@
+namespace KokkosKernelsGraphExample {
+
+template<typename OrdinalType, typename Device, typename MemoryTraits, typename SizeType>
+class StaticCrsGraph
```
Same as above. Why do we have this here ?

srajama1 commented:
@william76 What is happening with this PR ?

william76 commented:
@srajama1 I think you guys were wanting me to switch some things up to use the stuff @brian-kelley was putting into #466 but I didn't have time to do that prior to the end of the FY.  I'm not sure when I'll have time available to finish that refactor up though... :(

srajama1 commented:
I wish we wrapped up all the things for the FY :(. We will take care of it.

william76 commented:
@ndellingwood @srajama1 I worked with @brian-kelley on Monday to clean up some things from the original state of the example. He was extremely helpful in pointing me in the right direction on a few things. We were able to get rid of the Matrix and Graph classes I had copied over from `perf_test` and get the example more self contained using as much as possible from the Kokkos-Kernels and Kokkos libraries proper and to make sure we're loading the MM graphs properly, etc.

The example basically demonstrates:
1. Using D2GC 
2. How to get the colors out of the Kernel Handle.
3. How to run the correctness validation routine `KokkosGraph::Impl::graph_verify_distance2_color()`.
4. How to save a [GraphViz][2] `.dot` file using the `dump_graphviz()` method in `GraphColorDistance2Handle`.
5. How to print out a histogram of the colors using `KokkosGraph::Impl::graph_print_distance2_color_histogram()`

(3) and (5) are in `KokkosGraph_Distance2Color_impl.hpp` file so they aren't necessarily part of the "public" API for D2GC but they're handy so I thought it's worth keeping them in the example.

I've also cleaned up the output that gets printed and also got rid of the Parameters.hpp file so now the example is now fully self-contained in one file.

I added a README.md file to the `example/graph` directory which describes what the example is doing, etc.  If you want to look at the rendered markdown it's probably easiest to [have a look at it on my fork/branch][1].

Let me know if this is what you're looking for or if there are any other adjustments you're interested in making.

[1]: https://github.com/william76/kokkos-kernels/tree/D2GC-Example/example/graph
[2]: https://www.graphviz.org/

srajama1 commented:
@william76 This is exactly what I wanted. Thanks for taking the time and doing it.

@brian-kelley : Can you review this PR , please ?

brian-kelley commented on example/graph/README.md:
```diff
@@ -0,0 +1,119 @@
+Distance-2 Graph Coloring is VALID
+
+Distance-2 Color Histogram:
+0 1767 1737 1774 1748 1619 1610 1596 1494 1467 1395 1358 1288 1272 1188 1161 1082 1050 1016 962 ... ... ... 30 26 22 21 16 18 13 13 12 8 8 7 6 5 5 4 3 2 2 2
```
@william76 Does the histogram dump start at color 0? It would be nice to start at color 1 since no vertex will ever be color 0. Maybe just print out a subview from 1 to the end.

brian-kelley replied:
@william76 Everything else is good.

william76 replied:
@brian-kelley How do I extract a sub-view for elements 1..N-1 of a view?

brian-kelley replied:
@william76 This would be from 1 to the end, so ``Kokkos::subview(histogram, std::make_pair(1, histogram.extent(0)))``.

william76 replied:
@brian-kelley hmmm, I changed my print function to do this:
```C++
auto histogram_slice = Kokkos::subview(histogram, std::make_pair(1,histogram.extent(0)));
```
but this doesn't compile:
```
In file included from /Users/wcmclen/dev/kk-dev/kk-test/darwin-Serial-OpenMP/kokkos/install/include/Kokkos_View.hpp:490:0,
                 from /Users/wcmclen/dev/kk-dev/kk-test/darwin-Serial-OpenMP/kokkos/install/include/Kokkos_Parallel.hpp:52,
                 from /Users/wcmclen/dev/kk-dev/kk-test/darwin-Serial-OpenMP/kokkos/install/include/Kokkos_Serial.hpp:55,
                 from /Users/wcmclen/dev/kk-dev/kk-test/darwin-Serial-OpenMP/kokkos/install/include/Kokkos_Core.hpp:53,
                 from /Users/wcmclen/dev/kk-dev/source/KokkosKernels/example/graph/KokkosKernels_Example_Distance2GraphColor.cpp:51:
/Users/wcmclen/dev/kk-dev/kk-test/darwin-Serial-OpenMP/kokkos/install/include/impl/Kokkos_ViewMapping.hpp: In instantiation of ‘struct Kokkos::Impl::is_integral_extent<0, std::pair<int, long unsigned int> >’:
/Users/wcmclen/dev/kk-dev/kk-test/darwin-Serial-OpenMP/kokkos/install/include/impl/Kokkos_ViewMapping.hpp:3231:12:   required from ‘class Kokkos::Impl::ViewMapping<void, Kokkos::ViewTraits<int*, Kokkos::HostSpace>, std::pair<int, long unsigned int> >’
/Users/wcmclen/dev/kk-dev/kk-test/darwin-Serial-OpenMP/kokkos/install/include/Kokkos_View.hpp:2487:1:   required by substitution of ‘template<class D, class ... P, class ... Args> typename Kokkos::Impl::ViewMapping<void, Kokkos::ViewTraits<DataType, Properties ...>, Args ...>::type Kokkos::subview(const Kokkos::View<D, P ...>&, Args ...) [with D = int*; P = {Kokkos::HostSpace}; Args = {std::pair<int, long unsigned int>}]’
/Users/wcmclen/dev/kk-dev/kk-test/darwin-Serial-OpenMP/install/include/KokkosGraph_Distance2Color_impl.hpp:527:39:   required from ‘void KokkosGraph::Impl::GraphColorDistance2<HandleType, lno_row_view_t_, lno_nnz_view_t_, clno_row_view_t_, clno_nnz_view_t_>::print_color_histogram() [with HandleType = KokkosGraph::GraphColorDistance2Handle<const int, const int, const int, Kokkos::OpenMP, Kokkos::HostSpace, Kokkos::HostSpace>; lno_row_view_t_ = Kokkos::View<const int*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::OpenMP, Kokkos::HostSpace>, Kokkos::MemoryTraits<0> >; lno_nnz_view_t_ = Kokkos::View<int*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::OpenMP, Kokkos::HostSpace>, Kokkos::MemoryTraits<0> >; clno_row_view_t_ = Kokkos::View<const int*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::OpenMP, Kokkos::HostSpace>, Kokkos::MemoryTraits<0> >; clno_nnz_view_t_ = Kokkos::View<int*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::OpenMP, Kokkos::HostSpace>, Kokkos::MemoryTraits<0> >]’
/Users/wcmclen/dev/kk-dev/kk-test/darwin-Serial-OpenMP/install/include/KokkosGraph_Distance2Color_impl.hpp:1795:9:   required from ‘void KokkosGraph::Impl::graph_print_distance2_color_histogram(KernelHandle*, typename KernelHandle::nnz_lno_t, typename KernelHandle::nnz_lno_t, lno_row_view_t_, lno_nnz_view_t_, lno_col_view_t_, lno_colnnz_view_t_, bool) [with KernelHandle = KokkosKernels::Experimental::KokkosKernelsHandle<int, int, double, Kokkos::OpenMP, Kokkos::HostSpace, Kokkos::HostSpace>; lno_row_view_t_ = Kokkos::View<const int*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::OpenMP, Kokkos::HostSpace>, Kokkos::MemoryTraits<0> >; lno_nnz_view_t_ = Kokkos::View<int*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::OpenMP, Kokkos::HostSpace>, Kokkos::MemoryTraits<0> >; lno_col_view_t_ = Kokkos::View<const int*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::OpenMP, Kokkos::HostSpace>, Kokkos::MemoryTraits<0> >; lno_colnnz_view_t_ = Kokkos::View<int*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::OpenMP, Kokkos::HostSpace>, Kokkos::MemoryTraits<0> >; typename KernelHandle::nnz_lno_t = int]’
/Users/wcmclen/dev/kk-dev/source/KokkosKernels/example/graph/KokkosKernels_Example_Distance2GraphColor.cpp:454:65:   required from ‘void KokkosKernels::Example::run_example(CrsGraph_type, DataType, KokkosKernels::Example::Parameters) [with ExecSpace = Kokkos::OpenMP; DataType = int; CrsGraph_type = Kokkos::StaticCrsGraph<int, Kokkos::LayoutLeft, Kokkos::OpenMP, int, void>; TempMemSpace = Kokkos::HostSpace; PersistentMemSpace = Kokkos::HostSpace]’
/Users/wcmclen/dev/kk-dev/source/KokkosKernels/example/graph/KokkosKernels_Example_Distance2GraphColor.cpp:505:9:   required from ‘void KokkosKernels::Example::driver(KokkosKernels::Example::Parameters) [with size_type = int; lno_type = int; exec_space = Kokkos::OpenMP; hbm_mem_space = Kokkos::HostSpace]’
/Users/wcmclen/dev/kk-dev/source/KokkosKernels/example/graph/KokkosKernels_Example_Distance2GraphColor.cpp:544:119:   required from here
/Users/wcmclen/dev/kk-dev/kk-test/darwin-Serial-OpenMP/kokkos/install/include/impl/Kokkos_ViewMapping.hpp:315:3: error: static assertion failed: subview argument must be either integral or integral extent
   static_assert( value ||
   ^~~~~~~~~~~~~
```
Any of that make sense to you?

brian-kelley replied:
@william76 I think Kokkos is confused because the elements of the pair are different. Could you try casting the 1 to ``size_type``? I think that'll work...

william76 replied:
@brian-kelley well, that worked.  :man_facepalming:
```C++
auto histogram_slice = Kokkos::subview(histogram, std::make_pair((size_t)1, histogram.extent(0)));
```
It looks like it does the right thing. I'll do a quick rebuild and if all looks ok I'll push the change up.

william76 replied:
@brian-kelley I pushed the change in.

brian-kelley commented:
@srajama1 I think this is ready to go in, can you merge it?

srajama1 commented:
Merging it in @brian-kelley @william76 : Thank you, both !

william76 commented:
@srajama1 Just checking in... I saw this was still on my list of open PR's. Are you planning to merge and close on a later date?

srajama1 commented:
I didn't click the green button, sorry ! Doing it now.

william76 commented:
No worries... was wondering if it was that or if you were waiting until after y'all got past a sprint or something ;)
Thanks!
